app/controllers/integrations/github_controller.rb

```diff
@@ -20,21 +21,35 @@ def valid_signature?
   end
 
   def valid_payload?
-    request.headers['X-Github-Event'] == 'push'
+    webhook_event && ACCEPTED_EVENTS.include?(webhook_event.event_type)
   end
 
   def commit
-    params[:after]
+    webhook_event.sha
   end
 
   def branch
-    # Github returns full ref e.g. refs/heads/...
-    params[:ref][/refs\/heads\/(.+)/, 1]
+    webhook_event.branch
   end
 
   private
 
+  def webhook_event
+    return @webhook_event if defined?(@webhook_event)
+
+    @webhook_event = begin
+      case request.headers['X-Github-Event']
+      when 'push'
+        Changeset::CodePush.new(project.repo_name, params)
+      when 'pull_request'
+        Changeset::PullRequest.find(project.repo_name, params[:number])
+      when 'issue_comment'
+        Changeset::IssueComment.new(project.repo_name, params)
+      end
+    end
+  end
+
   def service_type
-    'code'
+    webhook_event && webhook_event.service_type
```

Comment:
Maybe we raise in valid_payload or something so we don't have to worry about there being a nil webhook_event